### PR TITLE
Add cross region resource management smoke tests

### PR DESCRIPTION
**Pull request by A-Hilaly**
Issue #81

Description of changes:
- Add AWS_REGION and AWS_PROFILE arguments to `sns_topic_exists`
- Add `sns/cross-region.sh` script to test cross region feature

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by A-Hilaly** [comment]
Tested with the changes made in #499

File: test/e2e/sns/cross-region.sh [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+
+topic_name="ack-test-smoke-$service_name"
+resource_name="topics/$topic_name"
+topic_arn_override_region="arn:aws:sns:$AWS_REGION_OVERRIDE:$AWS_ACCOUNT_ID:$topic_name"
+topic_arn_default_region="arn:aws:sns:$AWS_REGION_DEFAULT:$AWS_ACCOUNT_ID:$topic_name"
+
+
+# PRE-CHECKS
+if sns_topic_exists "$topic_arn_override_region" "$AWS_REGION_OVERRIDE"; then
+    echo "FAIL: expected $topic_name to not exist in SNS. Did previous test run cleanup?"
```
Comment (jaypipes) [inline test/e2e/sns/cross-region.sh:29]: Please include the region in this error message :+1: 

I was super confused about this before I realized what was going on :)

```
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ make kind-test SERVICE=sns AWS_ROLE_ARN=$ROLE_ARN
checking AWS credentials ... ok.
creating kind cluster ack-test-e59251b8-eb1e6e2a ... ok.
building aws-controllers-k8s:sns-2b66606 docker image ... ok.
loading the images into the cluster ... ok.
loading CRD manifests for sns into the cluster ... ok.
loading RBAC manifests for sns into the cluster ... ok.
loading service controller Deployment for sns into the cluster ...ok.
generating AWS temporary credentials and adding to env vars map ... ok.
======================================================================================================
To poke around your test cluster manually:
export KUBECONFIG=/home/jaypipes/go/src/github.com/aws/aws-controllers-k8s/scripts/lib/../../build/tmp-ack-test-e59251b8-eb1e6e2a/kubeconfig
kubectl get pods -A
======================================================================================================
testing Helm release for sns for release version 2b66606.
Building release artifacts for sns-2b66606
Generating custom resource definitions for sns
Generating RBAC manifests for sns
namespace/ack-system-test-helm created
installing the helm chart for ack-sns-controller in namespace ack-system-test-helm ... ok.
uninstalling the helm chart for ack-sns-controller in namespace ack-system-test-helm ... ok.
FAIL: expected ack-test-smoke-sns to not exist in SNS. Did previous test run cleanup?
To resume test with the same cluster use: "-c /home/jaypipes/go/src/github.com/aws/aws-controllers-k8s/scripts/lib/../../build/tmp-ack-test-e59251b8-eb1e6e2a"
make: *** [Makefile:51: kind-test] Error 1
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ daws sns list-topics
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ daws sns list-topics --region=eu-west-3
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ daws sns list-topics --region=eu-central-1
TOPICS	arn:aws:sns:eu-central-1:750630568209:ack-test-smoke-sns
```

File: test/e2e/sns/cross-region.sh [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    exit 1
+fi
+
+if sns_topic_exists "$topic_arn_default_region" "$AWS_REGION_DEFAULT"; then
+    echo "FAIL: expected $topic_name to not exist in SNS. Did previous test run cleanup?"
```
Comment (jaypipes) [inline test/e2e/sns/cross-region.sh:34]: ditto.

**Review by jaypipes — COMMENTED** [review]
This looks fine, but I'm getting failures when running it locally:

```
jaypipes@thelio:~/go/src/github.com/aws/aws-controllers-k8s$ make kind-test SERVICE=sns AWS_ROLE_ARN=$ROLE_ARN
checking AWS credentials ... ok.
creating kind cluster ack-test-ab89e97a-f525fdad ... ok.
building aws-controllers-k8s:sns-2b66606 docker image ... ok.
loading the images into the cluster ... ok.
loading CRD manifests for sns into the cluster ... ok.
loading RBAC manifests for sns into the cluster ... ok.
loading service controller Deployment for sns into the cluster ...ok.
generating AWS temporary credentials and adding to env vars map ... ok.
======================================================================================================
To poke around your test cluster manually:
export KUBECONFIG=/home/jaypipes/go/src/github.com/aws/aws-controllers-k8s/scripts/lib/../../build/tmp-ack-test-ab89e97a-f525fdad/kubeconfig
kubectl get pods -A
======================================================================================================
testing Helm release for sns for release version 2b66606.
Building release artifacts for sns-2b66606
Generating custom resource definitions for sns
Generating RBAC manifests for sns
namespace/ack-system-test-helm created
installing the helm chart for ack-sns-controller in namespace ack-system-test-helm ... ok.
uninstalling the helm chart for ack-sns-controller in namespace ack-system-test-helm ... ok.
namespace/testing-30190 created
topic.sns.services.k8s.aws/ack-test-smoke-sns created
topic.sns.services.k8s.aws "ack-test-smoke-sns" deleted
topic.sns.services.k8s.aws/ack-test-smoke-sns created
FAIL: expected ack-test-smoke-sns to have been created in SNS in region eu-west-3
2020-11-13T15:22:08.355Z	INFO	controller-runtime.metrics	metrics server is starting to listen	{"addr": "0.0.0.0:8080"}
2020-11-13T15:22:08.355Z	INFO	setup	initializing service controller	{"aws.service": "sns"}
2020-11-13T15:22:08.358Z	INFO	setup	starting manager	{"aws.service": "sns"}
2020-11-13T15:22:08.359Z	INFO	controller-runtime.manager	starting metrics server	{"path": "/metrics"}
2020-11-13T15:22:08.359Z	INFO	controller-runtime.controller	Starting EventSource	{"controller": "platformapplication", "source": "kind source: /, Kind="}
2020-11-13T15:22:08.359Z	INFO	controller-runtime.controller	Starting EventSource	{"controller": "platformendpoint", "source": "kind source: /, Kind="}
2020-11-13T15:22:08.359Z	INFO	controller-runtime.controller	Starting EventSource	{"controller": "topic", "source": "kind source: /, Kind="}
2020-11-13T15:22:08.360Z	DEBUG	ackrt.cache.namespace	created namespace	{"name": "kube-node-lease"}
2020-11-13T15:22:08.360Z	DEBUG	ackrt.cache.namespace	created namespace	{"name": "default"}
2020-11-13T15:22:08.360Z	DEBUG	ackrt.cache.namespace	created namespace	{"name": "local-path-storage"}
2020-11-13T15:22:08.360Z	DEBUG	ackrt.cache.namespace	created namespace	{"name": "kube-node-lease"}
2020-11-13T15:22:08.360Z	DEBUG	ackrt.cache.namespace	created namespace	{"name": "default"}
2020-11-13T15:22:08.360Z	DEBUG	ackrt.cache.namespace	created namespace	{"name": "local-path-storage"}
2020-11-13T15:22:08.478Z	INFO	controller-runtime.controller	Starting Controller	{"controller": "topic"}
2020-11-13T15:22:08.478Z	INFO	controller-runtime.controller	Starting Controller	{"controller": "platformapplication"}
2020-11-13T15:22:08.554Z	DEBUG	ackrt.cache.namespace	created namespace	{"name": "kube-node-lease"}
2020-11-13T15:22:08.554Z	DEBUG	ackrt.cache.namespace	created namespace	{"name": "default"}
2020-11-13T15:22:08.554Z	DEBUG	ackrt.cache.namespace	created namespace	{"name": "local-path-storage"}
2020-11-13T15:22:08.659Z	INFO	controller-runtime.controller	Starting Controller	{"controller": "platformendpoint"}
2020-11-13T15:22:08.659Z	INFO	controller-runtime.controller	Starting workers	{"controller": "platformendpoint", "worker count": 1}
2020-11-13T15:22:08.678Z	INFO	controller-runtime.controller	Starting workers	{"controller": "platformapplication", "worker count": 1}
2020-11-13T15:22:08.678Z	INFO	controller-runtime.controller	Starting workers	{"controller": "topic", "worker count": 1}
2020-11-13T15:22:18.171Z	DEBUG	ackrt.cache.namespace	created namespace	{"name": "ack-system-test-helm"}
2020-11-13T15:22:18.171Z	DEBUG	ackrt.cache.namespace	created namespace	{"name": "ack-system-test-helm"}
2020-11-13T15:22:18.171Z	DEBUG	ackrt.cache.namespace	created namespace	{"name": "ack-system-test-helm"}
2020-11-13T15:22:22.807Z	DEBUG	ackrt.cache.namespace	created namespace	{"name": "testing-30190"}
2020-11-13T15:22:22.807Z	DEBUG	ackrt.cache.namespace	created namespace	{"name": "testing-30190"}
2020-11-13T15:22:22.807Z	DEBUG	ackrt.cache.namespace	created namespace	{"name": "testing-30190"}
2020-11-13T15:22:22.956Z	DEBUG	ackrt	starting reconciliation	{"kind": "Topic", "namespace": "testing-30190", "name": "ack-test-smoke-sns", "generation": 1, "account": "750630568209", "role": "", "region": "eu-central-1"}
2020-11-13T15:22:22.961Z	DEBUG	ackrt	marked resource as managed	{"kind": "Topic", "namespace": "testing-30190", "name": "ack-test-smoke-sns", "generation": 1}
2020-11-13T15:22:24.096Z	INFO	ackrt	created new resource	{"kind": "Topic", "namespace": "testing-30190", "name": "ack-test-smoke-sns", "generation": 1, "is_adopted": false, "arn": "arn:aws:sns:eu-central-1:750630568209:ack-test-smoke-sns"}
2020-11-13T15:22:24.100Z	DEBUG	ackrt	patched resource	{"kind": "Topic", "namespace": "testing-30190", "name": "ack-test-smoke-sns", "generation": 1}
2020-11-13T15:22:24.100Z	DEBUG	controller-runtime.controller	Successfully Reconciled	{"controller": "topic", "request": "testing-30190/ack-test-smoke-sns"}
2020-11-13T15:22:24.100Z	DEBUG	ackrt	starting reconciliation	{"kind": "Topic", "namespace": "testing-30190", "name": "ack-test-smoke-sns", "generation": 1, "account": "750630568209", "role": "", "region": "eu-central-1"}
2020-11-13T15:22:24.247Z	DEBUG	ackrt	desired resource state has changed	{"kind": "Topic", "namespace": "testing-30190", "name": "ack-test-smoke-sns", "generation": 1, "is_adopted": false, "diff": "\"Status.EffectiveDeliveryPolicy\"\n", "arn": "arn:aws:sns:eu-central-1:750630568209:ack-test-smoke-sns"}
2020-11-13T15:22:24.250Z	DEBUG	ackrt	patched resource	{"kind": "Topic", "namespace": "testing-30190", "name": "ack-test-smoke-sns", "generation": 1}
2020-11-13T15:22:24.251Z	ERROR	controller-runtime.controller	Reconciler error	{"controller": "topic", "request": "testing-30190/ack-test-smoke-sns", "error": "InvalidParameter: 2 validation error(s) found.\n- missing required field, SetTopicAttributesInput.AttributeName.\n- missing required field, SetTopicAttributesInput.TopicArn.\n"}
github.com/go-logr/zapr.(*zapLogger).Error
	/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:258
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:232
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:211
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.Until
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:90
2020-11-13T15:22:25.251Z	DEBUG	ackrt	starting reconciliation	{"kind": "Topic", "namespace": "testing-30190", "name": "ack-test-smoke-sns", "generation": 1, "account": "750630568209", "role": "", "region": "eu-central-1"}
2020-11-13T15:22:25.429Z	DEBUG	controller-runtime.controller	Successfully Reconciled	{"controller": "topic", "request": "testing-30190/ack-test-smoke-sns"}
2020-11-13T15:22:39.808Z	DEBUG	ackrt	starting reconciliation	{"kind": "Topic", "namespace": "testing-30190", "name": "ack-test-smoke-sns", "generation": 2, "account": "750630568209", "role": "", "region": "eu-central-1"}
2020-11-13T15:22:40.825Z	INFO	ackrt	deleted resource	{"kind": "Topic", "namespace": "testing-30190", "name": "ack-test-smoke-sns", "generation": 2}
2020-11-13T15:22:40.831Z	DEBUG	ackrt	removed resource from management	{"kind": "Topic", "namespace": "testing-30190", "name": "ack-test-smoke-sns", "generation": 2}
2020-11-13T15:22:40.831Z	DEBUG	controller-runtime.controller	Successfully Reconciled	{"controller": "topic", "request": "testing-30190/ack-test-smoke-sns"}
2020-11-13T15:22:40.831Z	DEBUG	controller-runtime.controller	Successfully Reconciled	{"controller": "topic", "request": "testing-30190/ack-test-smoke-sns"}
2020-11-13T15:22:57.819Z	DEBUG	ackrt	starting reconciliation	{"kind": "Topic", "namespace": "testing-30190", "name": "ack-test-smoke-sns", "generation": 1, "account": "750630568209", "role": "", "region": "eu-west-3"}
2020-11-13T15:22:57.822Z	DEBUG	ackrt	marked resource as managed	{"kind": "Topic", "namespace": "testing-30190", "name": "ack-test-smoke-sns", "generation": 1}
2020-11-13T15:22:58.715Z	INFO	ackrt	created new resource	{"kind": "Topic", "namespace": "testing-30190", "name": "ack-test-smoke-sns", "generation": 1, "is_adopted": false, "arn": "arn:aws:sns:eu-central-1:750630568209:ack-test-smoke-sns"}
2020-11-13T15:22:58.719Z	DEBUG	ackrt	patched resource	{"kind": "Topic", "namespace": "testing-30190", "name": "ack-test-smoke-sns", "generation": 1}
2020-11-13T15:22:58.719Z	DEBUG	controller-runtime.controller	Successfully Reconciled	{"controller": "topic", "request": "testing-30190/ack-test-smoke-sns"}
2020-11-13T15:22:58.719Z	DEBUG	ackrt	starting reconciliation	{"kind": "Topic", "namespace": "testing-30190", "name": "ack-test-smoke-sns", "generation": 1, "account": "750630568209", "role": "", "region": "eu-west-3"}
2020-11-13T15:22:58.866Z	DEBUG	ackrt	desired resource state has changed	{"kind": "Topic", "namespace": "testing-30190", "name": "ack-test-smoke-sns", "generation": 1, "is_adopted": false, "diff": "\"Status.EffectiveDeliveryPolicy\"\n", "arn": "arn:aws:sns:eu-central-1:750630568209:ack-test-smoke-sns"}
2020-11-13T15:22:58.870Z	DEBUG	ackrt	patched resource	{"kind": "Topic", "namespace": "testing-30190", "name": "ack-test-smoke-sns", "generation": 1}
2020-11-13T15:22:58.870Z	ERROR	controller-runtime.controller	Reconciler error	{"controller": "topic", "request": "testing-30190/ack-test-smoke-sns", "error": "InvalidParameter: 2 validation error(s) found.\n- missing required field, SetTopicAttributesInput.AttributeName.\n- missing required field, SetTopicAttributesInput.TopicArn.\n"}
github.com/go-logr/zapr.(*zapLogger).Error
	/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:258
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:232
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:211
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.Until
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:90
2020-11-13T15:22:59.870Z	DEBUG	ackrt	starting reconciliation	{"kind": "Topic", "namespace": "testing-30190", "name": "ack-test-smoke-sns", "generation": 1, "account": "750630568209", "role": "", "region": "eu-west-3"}
2020-11-13T15:23:00.033Z	DEBUG	controller-runtime.controller	Successfully Reconciled	{"controller": "topic", "request": "testing-30190/ack-test-smoke-sns"}
To resume test with the same cluster use: "-c /home/jaypipes/go/src/github.com/aws/aws-controllers-k8s/scripts/lib/../../build/tmp-ack-test-ab89e97a-f525fdad"
```

I am not sure, however, whether the failures are due to problems introduced in this patch's test. I think perhaps we are hiding a permissions error improperly in the SNS service controller...

File: test/e2e/sns/cross-region.sh (hunk 1 above)
Comment (jaypipes) [reply test/e2e/sns/cross-region.sh:29]: @A-Hilaly maybe I need to apply https://github.com/aws/aws-controllers-k8s/pull/499 before running this?

**Comment by A-Hilaly** [comment]
> topic.sns.services.k8s.aws/ack-test-smoke-sns created
topic.sns.services.k8s.aws "ack-test-smoke-sns" deleted
topic.sns.services.k8s.aws/ack-test-smoke-sns created
FAIL: expected ack-test-smoke-sns to have been created in SNS in region eu-west-3

@jaypipes This error should be fixed after #499 merge, if it's the right solution of course

File: test/e2e/sns/cross-region.sh (hunk 2 above)
Comment (A-Hilaly) [reply test/e2e/sns/cross-region.sh:34]: done

File: test/e2e/sns/cross-region.sh (hunk 1 above)
Comment (A-Hilaly) [reply test/e2e/sns/cross-region.sh:29]: > Please include the region in this error message 👍

Done. Sorry for the confusion. My bad

Comment (A-Hilaly) [reply test/e2e/sns/cross-region.sh:29]: > @A-Hilaly maybe I need to apply #499 before running this?

@jaypipes After cleaning up the test objects (resource in k8s and AWS), and #499 everything should be OK.

**Comment by A-Hilaly** [comment]
> 2020-11-13T15:22:24.251Z	ERROR	controller-runtime.controller	Reconciler error	{"controller": "topic", "request": "testing-30190/ack-test-smoke-sns", "error": "InvalidParameter: 2 validation error(s) found.\n- missing required field, SetTopicAttributesInput.AttributeName.\n- missing required field, SetTopicAttributesInput.TopicArn.\n"}
github.com/go-logr/zapr.(*zapLogger).Error
	/go/pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:128
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:258
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:232
sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).worker
	/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.6.0/pkg/internal/controller/controller.go:211
k8s.io/apimachinery/pkg/util/wait.BackoffUntil.func1
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:155
k8s.io/apimachinery/pkg/util/wait.BackoffUntil
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:156
k8s.io/apimachinery/pkg/util/wait.JitterUntil
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:133
k8s.io/apimachinery/pkg/util/wait.Until
	/go/pkg/mod/k8s.io/apimachinery@v0.18.6/pkg/util/wait/wait.go:90

I believe this is happening because SetAttributes request doesn't set `TopicARN` and `AttributeName`( https://github.com/aws/aws-controllers-k8s/commit/22b1c9302630006e42bea22946a83e30abc9b842#diff-105c06cb396147de7ebbf80f44e107515501892f0bc8303a12755c748e5bf92cR248-R253 )

File: test/e2e/sns/cross-region.sh (hunk 1 above)
Comment (jaypipes) [reply test/e2e/sns/cross-region.sh:29]: > > Please include the region in this error message +1
> 
> Done. Sorry for the confusion. My bad

@A-Hilaly I don't see that you updated the message?

Comment (A-Hilaly) [reply test/e2e/sns/cross-region.sh]: @jaypipes Apologies, i might have forgotten to push my changes. Messages fixed and PR rebased.

```
./test/e2e/sns/cross-region.sh
namespace/testing-3897 created
topic.sns.services.k8s.aws/ack-test-smoke-sns created
topic.sns.services.k8s.aws "ack-test-smoke-sns" deleted
topic.sns.services.k8s.aws/ack-test-smoke-sns created
topic.sns.services.k8s.aws "ack-test-smoke-sns" deleted
namespace "testing-3897" deleted

```

**Review by jaypipes — APPROVED** [review]
Merci, monsieur :)

**Pull request merged**